Fixing tabs in copy-paste (#219)

sblundy opened the pull request:
Adding tab characters to the Grid as an actual tab - rather than a space - followed by spacer characters. The spacer characters have the WIDE_CHAR_SPACER flag so the selection code will *not* pass them to the clipboard. This means `Term::string_from_selection()` will filter out the spacers and just return the tabs.

jwilm commented on src/term/mod.rs:
```diff
@@ -1411,23 +1411,33 @@ impl ansi::Handler for Term {
+            {
+                let cell = &mut self.grid[&self.cursor.point];
+                *cell = self.cursor.template;
+                cell.c = self.cursor.charsets[self.active_charset].map('\t');
```
If `cell.c` is going to be a `'\t'`, I think the rendered needs to be updated to handle it like a `' '`

sblundy replied:
Ok, I think I have the copy code referencing `self.tabs`. Two questions. 

1. `self.tabs` looks like it could use an abstraction around lookup by column because everything else is also by column. Mind if I add one?
2. The renderer is rendering tabs as 0 height & width glyph, like a space. So, at least on my box, it's being handled fine. Maybe leave the renderer alone until there's a problem?

sblundy commented:
I changed the copy code to use self.tabs and I updated the renderer to render '\t' as a space. Tabs are also being skipped in the same way spaces are. Also, I went ahead and wrapped self.tabs to they're indexed by Columns.

chrisduerr commented on src/term/mod.rs:
```diff
@@ -908,10 +907,23 @@ impl Term {
+                            } else if cell.c == ' ' {
+                                continue;
+                            }
+                        }
```
It would probably be good to document this block because it's not obvious that it skips whitespaces until the next tabstop.

sblundy replied:
@chrisduerr If you mean the way a tab can be partially selected as if it were a bunch of spaces, I totally agree. I've been working on a fix that will have to be a separate PR.

chrisduerr replied:
Yeah that's what I'm talking about.

chrisduerr commented:
I've fixed one issue with tab selection in my personal branch of this feature:
https://github.com/chrisduerr/alacritty/tree/copy-tabs

I've thought about properly implementing selection of tabs so the whole tab would be selected instead of just a single one, however since the scrollback branch introduces some major changes, this doesn't make sense at this point.

I'd recommend waiting with this PR until scrollback is merged, then I fix this so it acts in line with other terminal emulators.
